src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,571 @@
+            .await
+            .expect("Failed to start proxy");
+
+        let (mut client, _conn, _client_addr) = connect(addr).await;
+
+        let (_parts, body) = send_empty_request(&mut client).await;
+
+        assert_eq!(body.as_ref(), b"Hello, world!");
+    }
+
+    #[tokio::test]
+    async fn server_receives_decoded_client_address() {
+        let (_handle, _server_task, server_addr) = start_server(true, true).await;
+
+        let addr = start_proxy(server_addr, true)
+            .await
+            .expect("Failed to start proxy");
+
+        let (mut client, _conn, client_addr) = connect(addr).await;
+
+        let (_parts, body) = send_empty_request(&mut client).await;
+        let body_str = String::from_utf8(body.to_vec()).expect("Response body is not valid UTF-8");
+
+        assert_eq!(body_str, format!("for={}", client_addr));
+    }
+
+    #[tokio::test]
+    async fn not_parsing_when_header_present_fails() {
+        let (_handle, _server_task, server_addr) = start_server(false, false).await;
+
+        let addr = start_proxy(server_addr, true)
+            .await
+            .expect("Failed to start proxy");
+
+        let (mut client, _conn, _client_addr) = connect(addr).await;
+
+        let (_parts, body) = send_empty_request(&mut client).await;
+
+        assert!(body.as_ref() != b"Hello, world!");
+    }
+
+    #[tokio::test]
+    async fn parsing_when_header_not_present_fails() {
+        let (_handle, _server_task, server_addr) = start_server(true, false).await;
+
+        let addr = start_proxy(server_addr, false)
+            .await
+            .expect("Failed to start proxy");
+
+        let (mut client, _conn, _client_addr) = connect(addr).await;
+
+        match client
+            .ready()
+            .await
+            .unwrap()
+            .call(Request::new(Body::empty()))
+            .await
+        {
+            Ok(_) => panic!("Should have failed"),
+            Err(e) => {
+                if e.is_incomplete_message() {
+                } else {
+                    panic!("Received unexpected error");
+                }
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn test_shutdown() {
+        let (handle, _server_task, server_addr) = start_server(true, false).await;
+
+        let addr = start_proxy(server_addr, true)
+            .await
+            .expect("Failed to start proxy");
+
+        let (mut client, conn, _client_addr) = connect(addr).await;
+
+        handle.shutdown();
+
+        let response_future_result = client
+            .ready()
+            .await
+            .unwrap()
+            .call(Request::new(Body::empty()))
+            .await;
+
+        assert!(response_future_result.is_err());
+
+        // Connection task should finish soon.
+        let _ = timeout(Duration::from_secs(1), conn).await.unwrap();
+    }
+
+    pub(crate) async fn forward_ip_handler(req: Request<Body>) -> Response<Body> {
```

Comment:
Should care about the whole request, not just the body, we can't mangle other headers, or the server and services will break.

Reply:
We also can't mangle trailers for gRPC.